### ISSUE TITLE
[clang 18] x86-64 miscompile with boolean conditions

### ISSUE THREAD

**Issue by liamwhite** [llvm/llvm-project#108879]
The following C++ code ([minimized from from this report](https://gitlab.com/inkscape/inkscape/-/merge_requests/6696))
```cpp
struct ToolBase {};
struct CanvasEvent {};
struct Desktop {
    ToolBase* tool;
};

bool start_root_handler(ToolBase* tool, const CanvasEvent& event);
bool tool_handler(const CanvasEvent& event, Desktop* desktop) {
    if (auto const tool = desktop->tool) {
        return start_root_handler(tool, event);
    }
    return false;
}
```
produces broken compiled output with clang 18.1.x ([compiler explorer](https://godbolt.org/z/dco6hfs8f)):
```x86asm
tool_handler(CanvasEvent const&, Desktop*):
        push    rbx
        mov     rbx, qword ptr [rsi]
        test    rbx, rbx
        je      .LBB0_1
        mov     rax, rdi
        mov     rdi, rbx
        mov     rsi, rax
        call    start_root_handler(ToolBase*, CanvasEvent const&)@PLT
        jmp     .LBB0_3
.LBB0_1:
.LBB0_3:
        test    rbx, rbx
        setne   cl
        and     cl, al
        mov     eax, ecx
        pop     rbx
        ret
```

The issue occurs on these lines
```x86asm
        setne   cl
        and     cl, al
        mov     eax, ecx
```
Only the low byte of the c register is clobbered by `setne   cl`. However, the optimizer seemingly assumes that the entire `c` register has been clobbered, and ends up moving garbage/undefined 32-bit data in the `mov     eax, ecx` line.
```
(gdb) si
(gdb) p	$ecx
$5 = 2069502208
(gdb) p	$eax
$6 = 2069502208
(gdb) p	$cl
$7 = 0
(gdb) p	$al
$8 = 0
```
This results in the function returning a nonsense non-boolean value which breaks the functions that call it.

**Comment by liamwhite** [comment]
I did not verify with a debugger but this also appears to be broken on arm64:
```armasm
tool_handler(CanvasEvent const&, Desktop*):
        stp     x29, x30, [sp, #-32]!
        str     x19, [sp, #16]
        mov     x29, sp
        ldr     x19, [x1]
        cbz     x19, .LBB0_2
        mov     x1, x0
        mov     x0, x19
        bl      start_root_handler(ToolBase*, CanvasEvent const&)
        b       .LBB0_3
.LBB0_2:
.LBB0_3:
        cmp     x19, #0
        cset    w8, ne
        and     w0, w8, w0
        ldr     x19, [sp, #16]
        ldp     x29, x30, [sp], #32
        ret
```
After a jump to .LBB0_2, the value of x0/w0 has not been converted to a boolean and so this instruction sequence which is intended to produce a boolean output into w0 does not make sense:
```armasm
        cset    w8, ne
        and     w0, w8, w0
```

**Comment by dzaima** [comment]
Is this actually incorrect? It should be fine for bits 8..64 to be non-zero of a `bool` return value, it's the responsibility of the caller to only check the bottom 8: https://godbolt.org/z/b3hWWGsce

The arm64 example is less trivial, but in the path of unset `w0` the `cset    w8, ne` should always be setting `w8` to `0`, and `0 & garbage == 0`.

**Comment by liamwhite** [comment]
On further investigation, that seems correct, at least for SysV ABI. The underlying issue in the original example is a function returning C++ bool but was type punned to a function that expects to return a `gboolean` (i.e. an int). I will close this.
